lib/Doctrine/Migrations/MigrationTableManager.php

```diff
@@ -0,0 +1,224 @@
+            }
+        }
+
+        return $this->migrationTableCreated;
+    }
+
+    public function isMigrationTableUpToDate() : bool
+    {
+        if ($this->migrationTableUpToDate === null) {
+            $table = $this->getCurrentTable();
+
+            $this->migrationTableUpToDate = true;
+
+            foreach ($this->getColumns() as $columnName => $column) {
+                if ($table->hasColumn($columnName) === false) {
+                    $this->migrationTableUpToDate = false;
+
+                    break;
+                }
+            }
+        }
+
+        return $this->migrationTableUpToDate;
+    }
+
+    public function createMigrationTable() : bool
+    {
+        $this->configuration->validate();
+
+        if ($this->configuration->isDryRun()) {
+            return false;
+        }
+
+        if ($this->isMigrationTableCreated()) {
+            if ($this->isMigrationTableUpToDate() === false) {
+                $this->updateMigrationTable();
+
+                $this->migrationTableUpToDate = true;
+
+                return true;
+            }
+
+            return false;
+        }
+
+        $table = $this->getToTable();
+
+        $this->schemaManager->createTable($table);
+
+        $this->migrationTableCreated = true;
+
+        return true;
+    }
+
+    public function getMigrationsColumn() : Column
+    {
+        return new Column(
+            $this->columnName,
+            Type::getType(self::MIGRATION_COLUMN_TYPE),
+            ['length' => $this->columnLength]
+        );
+    }
+
+    public function getExecutedAtColumn() : Column
+    {
+        return new Column(
+            $this->executedAtColumnName,
+            Type::getType('datetime'),
+            ['notnull' => false]
+        );
+    }
+
+    /**
+     * @param Table[] $tables
+     */
+    protected function createSchema(array $tables) : Schema
+    {
+        return new Schema($tables);
+    }
+
+    /**
+     * @return Column[]
+     */
+    private function getColumns() : array
+    {
+        return [
+            $this->columnName           => $this->getMigrationsColumn(),
+            $this->executedAtColumnName => $this->getExecutedAtColumn(),
+        ];
+    }
+
+    private function getCurrentTable() : Table
+    {
+        return $this->schemaManager->listTableDetails($this->tableName);
+    }
+
+    private function getFromTable() : Table
+    {
+        $currentTable = $this->getCurrentTable();
+
+        $columnNames = array_keys($this->getColumns());
+
+        $table = new Table($this->tableName, $currentTable->getColumns());
+        $table->setPrimaryKey([$this->columnName]);
+
+        // remove columns from the table definition that we don't care about
+        // so we don't try to drop those columns
+        foreach ($table->getColumns() as $column) {
+            if (in_array($column->getName(), $columnNames, true)) {
+                continue;
+            }
+
+            $table->dropColumn($column->getName());
```

Comment:
If I understood correctly, then on the happy path we would not need to drop any columns from the definition - what do you think about modifying to

```diff
diff --git a/lib/Doctrine/Migrations/MigrationTableManager.php b/lib/Doctrine/Migrations/MigrationTableManager.php
index 175c2da..721116c 100644
--- a/lib/Doctrine/Migrations/MigrationTableManager.php
+++ b/lib/Doctrine/Migrations/MigrationTableManager.php
@@ -189,11 +189,9 @@ class MigrationTableManager
         // remove columns from the table definition that we don't care about
         // so we don't try to drop those columns
         foreach ($table->getColumns() as $column) {
-            if (in_array($column->getName(), $columnNames, true)) {
-                continue;
+            if (!in_array($column->getName(), $columnNames, true)) {
+                $table->dropColumn($column->getName());
             }
-
-            $table->dropColumn($column->getName());
         }

         return $table;
```

?

Reply:
That goes against early-exit rule in conditions/loops in Doctrine CS. :)